Add guideline for evaluating proposals (#466)
godotengine/godot-proposals: README.md

apart from bug reports.

Proposals are made by opening issues in this repository, which will
then be discussed with fellow Godot users and contributors. If a
proposal is considered good to implement, it will be approved by a
core developer.

Please note: making a proposal does not guarantee that your proposed
feature will be accepted into the engine. In fact, in most cases your
proposal will not be accepted. The Godot organization has very few
resources and relies almost completely on irregular contributors to
implement anything. Technical debt and software bloat accumulate

> quickly, as a result core developers have to be very selective about

Reviewer: ```suggestion
quickly; as a result, core developers have to be very selective about
```

Reviewer 2: Not fixed.

what features make it into the engine. For a deeper discussion on what

> features will be accepted see the section **How core developers

Reviewer: ```suggestion
features will be accepted, see the section **How core developers
```

Reviewer 2: Not fixed.

> evaluate proposals** below.

Reviewer 2: While this is good to mention, I think it might be a bit *too* negative.

It's also important to keep in mind that the idea behind the proposals / GIP workflow is to outline technical proposals that can be implemented by any contributor with the blessing of the core team -- so resources available to current core contributors are only part of the equation - it's fine to have an approved but implemented proposal taking dust for a while until a new contributor that needs it comes up to implement it.

Author: I agree. However, I also think that setting expectations is important. Do you think it could be rephrased in a better tone, or would you like to cut this paragraph entirely?

Reviewer 3: > Please note: making a proposal does not guarantee that your proposed
feature will be accepted into the engine.

That's the definition of a proposal: it may be accepted, or it may be not, I'd remove this. (again nobody is forcing anyone to do anything, even if one thinks that "someone has to do this").

> In fact, in most cases your proposal will not be accepted.

That's not a fact. Some might have a good eye on what features are suited for core. But I guess it's safe to say that the overall likelihood of proposed features being implemented is very low, but again if a person follows the evaluation guideline, the likelihood of a feature being implemented should be very high, unless the discretion of core developers takes higher priority. So I'd replace `your` with a more general statement using statistics, making it less personal. Something like this:
```
Bare in mind that most proposals may not end up being implemented directly in the engine.
```

That also tells how a proposed feature can be a good candidate for a plugin if a proposal is rejected for core, and becomes in alignment with `Technical debt and software bloat` statements.

I'm not yet sure whether the guideline is written for people *making* proposals, or for those *evaluating* proposals, but I guess both. If that's the case, any personal references such as `you` or `your` do not apply.

Reviewer 4: Nitpick: Bear not bare in mind.

Author: I have cut the paragraph entirely. Better to remain positive.

template to show how Godot is not currently meeting your needs and then
explain how your proposal will meet a particular need. 

> 4. Other users must express interest in your proposal for it to be considered.

Reviewer: Change the bullet point number to account for the removed bullet point above:

```suggestion
3. Other users must express interest in your proposal for it to be considered.
```

Reviewer 2: It still shows as `4.` for me.

not worth pursuing and you feel that it was wrongly closed, then feel free
to join the developer IRC group and have a more in-depth discussion with 
other core developers about the feature.

## How core developers evaluate proposals

The following is a list of considerations that core developers use when deciding
to accept, close, or leave a proposal open. It is intended to be useful for core
developers when considering proposals and for proposal-makers in drafting their
proposals.

> **1. Does the proposal comply with the rules?**

Reviewer: It may be better to use `###` headings instead of bold text, as some Markdown viewers don't break lines unless you add a blank line below.

Author: Done. :)
I used ``####`` to make the titles closer to normal size.

Read the proposal and check to see that it complies with the above-stated rules.
If it does not, close the proposal.

**2. How much support is the proposal receiving?**
Evaluate the amount of support the proposal is receiving. This is an ongoing
analysis. If a proposal receives little support at first, it may receive
additional support later on.

**3. Can this proposal be implemented with an addon?**

> Evaluate whether it is possible for the proposal to be implemented inan addon.

Reviewer: ```suggestion
Evaluate whether it is possible for the proposal to be implemented in an addon.
```

Reviewer 2: Not fixed.

> If it is possible for the proposal to be in an addon it is unlikely to be

Reviewer: ```suggestion
If it is possible for the proposal to be in an addon, it is unlikely to be
```

Reviewer 2: Not fixed.

accepted.

**4. Does this proposal benefit most users?**
Determine whether this proposal benefits all users, or just certain users.
For example, a feature that can only be used for 3D FPS games is less
likely to be accepted than a feature that benefits all 3D games.

**5. Can this proposal be implemented in a robust, general-purpose way?**
Determine whether the feature can be implemented in a robust way that benefits

> all use-cases. For example, many games use an inventory system, however,

Reviewer: ```suggestion
all use-cases. For example, many games use an inventory system. However,
```

Reviewer 2: Not fixed.

every game implements inventory differently. Accordingly, a proposal for an
inventory system will likely not be accepted because it would be impossible for
us to implement an inventory system that works for most users that need an
inventory in their game.

**6. Does this proposal help users overcome a limitation?**
Proposals that overcome a specific limitation are more likely to be accepted
than proposals that are just helpful. In short, need-to-have features will be
prioritized over nice-to-have features.

**7. How complex would the proposed feature be?**
A highly complex new feature involving substantial changes to core is less likely
to be accepted than a feature that can be contained within a single node, or a
group of nodes.

**8. Can the feature be worked around in script with a few lines?**

> If the feature is intended to save users a few lines of code it is unlikely

Reviewer: ```suggestion
If the feature is intended to save users a few lines of code, it is unlikely
```

to be accepted.

The above considerations are all balanced, no one is more important than another.
Core developers have discretion to weigh the factors as they see fit.

> In addition to the above guideline consider The graph in [this article](https://godotengine.org/article/will-your-contribution-be-merged-heres-how-tell) which outlines what core developers consider when evaluating PRs.

Reviewer: ```suggestion
In addition to the above guideline, consider the graph in [this article](https://godotengine.org/article/will-your-contribution-be-merged-heres-how-tell) which outlines what core developers consider when evaluating PRs.
```

Reviewer 2: Not fixed.

Reviewer 2: It's the only line which is not hard wrapped, might be worth fixing.
But overall it would be better to refer to https://docs.godotengine.org/en/latest/community/contributing/best_practices_for_engine_contributors.html than the less nuanced graph of that earlier blog post IMO.

Reviewer 3: Yes, I personally find the graph discouraging to say the least, talking about https://godotengine.org/article/will-your-contribution-be-merged-heres-how-tell.

While I do understand the rationale, the "We *refuse* by default" statement does not contribute to the open mindedness of open source software development, but again maybe I'm lacking experience to say that.

Else we get "proposals" like #486 with the opposite attitude of "I *propose* by default". 🙂

Author: Changed to point to the new doc.

not worth pursuing and you feel that it was wrongly closed, then feel free
to join the developer IRC group and have a more in-depth discussion with 
other core developers about the feature.

## How core developers evaluate proposals

The following is a list of considerations that core developers use when deciding
to accept, close, or leave a proposal open. It is intended to be useful for core
developers when considering proposals and for proposal-makers in drafting their
proposals.

#### 1. Does the proposal comply with the rules?
Read the proposal and check to see that it complies with the above-stated rules.
If it does not, close the proposal.

#### 2. How much support is the proposal receiving?
Evaluate the amount of support the proposal is receiving. This is an ongoing
analysis. If a proposal receives little support at first, it may receive
additional support later on.

> #### 3. Can this proposal be implemented with an addon?

Reviewer 2: This has to be weighed together with how useful/frequently needed the proposal would be. If the majority of users would benefit from it, it doesn't necessarily make sense to force it to be an addon.
We can refer to https://docs.godotengine.org/en/latest/community/contributing/best_practices_for_engine_contributors.html#the-problem-has-to-be-complex-or-frequent

In general, we can refer to the whole page as one of the guiding principles for reviewing proposals (just like we use it for reviewing PRs).

Reviewer 2: Actually many of the points that come below are covered by https://docs.godotengine.org/en/latest/community/contributing/best_practices_for_engine_contributors.html - I guess your PR served as inspiration for @reduz :)

We *might* want to reduce redundancy by referring to that docs page instead of restating everything, but at the same time it's likely good to have things written clearly in the readme.

Author: Yea, there is definitely a lot of overlap. But they are still addressing two distinct processes (PRs v. Proposals). My preference is to keep the two separate, as most of the proposals we receive are from people who do not contribute to the engine and that doc is likely to confuse them as it isn't clearly aimed at proposals.

Author: I have chosen to keep this section mostly as is. Changing to "less likely" instead of "unlikely". The purpose behind each section is to show an independent factor that must be weighed. All factors should be weighed against each other. IMO it is clearer to say that this factor lowers chances while other factors raise chances instead of mixing all the factors into each.
